Add `empty` as a parameter to DelimitedFieldMixin

sirosen opened the pull request:
`empty` allows more graceful support for non-string delimited data, like lists of ints.

`empty` is annotated on the DelimitedFieldMixin class but is only set on its subclasses. This technically means that a custom subclass of the mixin class could be defined by a third party and would be broken by this change, but this is acceptable.

resolves #823

sirosen committed:
Add `empty` as a parameter to DelimitedFieldMixin

`empty` allows more graceful support for non-string delimited data,
like lists of ints.

`empty` is annotated on the DelimitedFieldMixin class but is only set
on its subclasses. This technically means that a custom subclass of
the mixin class could be defined by a third party and would be broken
by this change, but this is acceptable.

resolves #823

lafrech commented on tests/test_core.py:
```diff
@@ -1104,6 +1099,32 @@ def test_delimited_tuple_passed_invalid_type(web_request, parser):
+    assert parsed["ids"] == [1, 0, 3]
+
+    data = schema.dump(parsed)
+    assert data["ids"] == "1,0,3"
```
Not a blocker at all but do we need to test `dump`?

sirosen replied:
I'm not overly attached to checking it...

It keeps the tests uniform with other delimited field tests, but I'd say it's not a valuable check in these cases. Why don't I remove it? No point keeping it if we agree that it's not doing anything.

lafrech replied:
I didn't realize other tests did this. I'm fine both ways.

sirosen replied:
I removed it. Once it was pointed out, it started to bother me. :sweat_smile:

lafrech reviewed (commented):
I'm happy with this implementation.

Before we finalize this, I'd be interested to see how OpenAPI deals with this. Not all webargs users do REST and use apispec, but it may be a valuable hint. Last time I checked, I didn't find much. Looks like a corner/advanced case.

lafrech commented:
While having a look at how to document this, I came to think that maybe we're doing this the wrong way.

The `DelimitedTuple` case is a hint that solving this at list level rather than nested schema level might be a bit wrong. Indeed, one could imagine a tuple with mixed strings and integers:

```python
    DelimitedLTuple(String(load_default=""), Integer(load_default=0)
```

What if we decided that `DelimitedList` passes `missing` to the nested schema rather than the "empty" value, then we'd just rely on nested schema default?

I'm not sure this would work as-is, and I'm done for tonight, but we should think about this before merging.

Also this would break the `DelimitedList(String)` use case because users would have to set a default explicitly, but maybe that's the way it is. 

```python
    DelimitedList(String(load_default=""))
```

sirosen commented:
I'm not sure I like the idea of requiring `load_default` for the current behavior. `DelimitedList(String())` parsing `"a,b,,c"` into `["a", "b", "", "c"]` seems to me like the most intuitive thing.

But you're also right that passing `missing` allows the nested field to decide how to handle this...

What about adding `empty=""` as the default, but documenting `empty=ma.missing` in the advanced docs as a way of getting more power out of the delimited fields? I think there's an argument in favor of the default being `missing` from a "purity" perspective, but that `""` is more practical because it's what more users will want.

lafrech commented:
I agree about the need for a practical default.

Yet the hybrid tuple case bothers me.

Note that I didn't check that passing missing actually works. It might silently hide empty items. I dunno, I didn't try anything.

Perhaps practicality and purity could be achieved by using subclasses or mixins to make the choice explicit without requiring an "empty" argument.

sirosen commented:
Just to quickly answer the question, passing `missing` works as you'd hope:
```python
>>> import webargs as wa
>>> f = wa.fields.DelimitedList(wa.fields.String(), empty=wa.missing)
>>> f.deserialize("a,,b")
['a', <marshmallow.missing>, 'b']
>>> f = wa.fields.DelimitedList(wa.fields.String(load_default="foo"), empty=wa.missing)
>>> f.deserialize("a,,b")
['a', 'foo', 'b']
```

Taking a little inspiration from pathlib's naming, I'm wondering if we'd be okay with this?
```python
# passes missing to subfields
PureDelimitedList
PureDelimitedTuple
# passes "" to subfields
DelimitedList
DelimitedTuple
```
with the implementation being that `empty_value` (longer name to be a little more explicit) is a classvar which each subclass can set, and which is `missing` on `DelimitedFieldMixin`.

I like, as a user of `pathlib`, that `PureX` vs `X` indicates clearly to the user that there's a subtle distinction between classes which you _might_ need to understand depending on your use-case.

sirosen commented:
I looked at my comment again and don't like `Pure` for naming. It looks _too much_ like `PurePath`, which has a different meaning from this.

If we do it with classes, maybe `UnfilteredDelimitedList`.

Also, it may make sense in v9 for tuple to default to missing while list defaults to `""`.

I'm still trying to find a good clean solve, so thoughts are welcome.

sirosen commented:
@lafrech, I have an idea for v9, but it would be backwards incompatible. Would you be okay with the following plan?

- set `empty_value` as a class variable in v8, equal to `""`, rework this PR to implement that
- in v9, `empty_value` will default to `missing`
- in v9, the default behavior of `DelimitedList` _if given no field as an argument_ is to insert a field `String(load_default="")`
- we document these changes and the rationale as more controlled handling of empty values

I would want to work more on this and maybe tinker with related ideas when we prepare v9, but this way we can close this issue out today and have a plan for the future. Having `DelimitedList()` produce a list of strings (including empty strings) satisfies my desire to make the easy case easy.

sirosen commented:
I refactored this and rebased. It now uses `empty_value` as a class attribute (no support for `empty=...` in the field constructor) and the changelog has been updated to list the change.

The implementation barely changes in this, but it will set us up for the desired change for 9.0 .

lafrech reviewed (approved):
Great.

It's nice that users can already opt-in for `missing` and in v9 they'll still be able to set `""` as default if they wish.
